File: java/src/main/java/structural/bridge/MainBridge.java
package structural.bridge;

interface Motor {
	public Integer getHorsepower();
}

class StockMotor implements Motor {
	@Override
	public Integer getHorsepower() {
		return 250;
	}
}

class RaceMotor implements Motor {
	@Override
	public Integer getHorsepower() {
		return 1000;
	}
}

abstract class Car {
	protected Motor motor;
	public Car(Motor motor) {
		this.motor = motor;
	}
	abstract public Motor getMotor();
	abstract public String getType();
}

class PassengerCar extends Car {
	public PassengerCar(Motor motor) {
		super(motor);
	}

	@Override
	public Motor getMotor() {
		return motor;
	}

	@Override
	public String getType() {
		return "Passenger car with " + this.motor.getHorsepower() + "hp";
	}
}

class SUV extends Car {
	public SUV(Motor motor) {
		super(motor);
	}

	@Override
	public Motor getMotor() {
		return motor;
	}
	
	@Override
	public String getType() {
		return "SUV with " + this.motor.getHorsepower() + "hp";
	}
}

public class MainBridge {
	public static void main(String[] args) {
		Car raceSUV = new SUV(new RaceMotor());
		System.out.println(raceSUV.getType());
		Car stockCar = new PassengerCar(new StockMotor());
		System.out.println(stockCar.getType());
	}
}

File: kotlin/src/main/kotlin/structural/bridge/MainBridge.kt
package structural.bridge

internal interface Motor {
    public fun getHorsepower(): Int
}

class StockMotor : Motor {
    override fun getHorsepower(): Int {
        return 250;
    }
}

class RaceMotor : Motor {
    override fun getHorsepower(): Int {
        return 1000;
    }
}

internal abstract class Car {
    protected var motor: Motor
        get() {return field}
        private set
    public constructor(motor: Motor) {
        this.motor = motor
    }
    abstract public fun getType(): String
}

internal class PassengerCar(motor: Motor) : Car(motor) {
    override fun getType(): String {
        return "Passenger car with " + motor.getHorsepower() + "hp"
    }
}

internal class SUV(motor: Motor) : Car(motor) {
    override fun getType(): String {
        return "SUV with " + motor.getHorsepower() + "hp"
    }
}

fun main(args: Array<String>) {
    val raceSUV: Car = SUV(RaceMotor())
    println(raceSUV.getType())
    val stockCar: Car = PassengerCar(StockMotor())
    println(stockCar.getType())
}